### PR TITLE
Fix temperature readout for Invensense IMUs

### PR DESCRIPTION
**Pull request by mhkabir**
This fixes the temperature readout for the MPU/ICM IMUs. The sensitivity parameters were originally incorrect. This is needed for enabling the Pixhawk 2.1 IMU heater and holding a temperature set-point.

### CODE REVIEW

**Comment by mhkabir** [comment]
TODO : other drivers also seem to have incorrect parameters.

**Comment by LorenzMeier** [comment]
Generally a great initiative, but we need to take into consideration what the side effect will be to boards that are already calibrated.

File: src/drivers/imu/mpu9250/mpu9250.cpp [hunk 1]
```diff
@@ -1434,7 +1434,12 @@ MPU9250::measure()
 	arb.scaling = _accel_range_scale;
 	arb.range_m_s2 = _accel_range_m_s2;
 
-	_last_temperature = (report.temp) / 361.0f + 35.0f;
+	// We calculate sensor temperature as :
+	// temperature_real = temperature_raw * temperature_sensitivity + zero_offset
+	//
+	// Sensitivity (in degC/LSB) is from the Invensense datasheet for the specific sensor
+	// Zero offset (in degC) varies from part to part, and this value is a good default
+	_last_temperature = (report.temp) / 340.0f + 21.0f;
```
Comment (ksschwabe) [inline src/drivers/imu/mpu9250/mpu9250.cpp]: Where do these numbers come from? Is the temp sensitivity not 333.87LSB/degC?
mpu9250_temp_sensitivity

File: src/drivers/imu/mpu6000/mpu6000.cpp [hunk 2]
```diff
@@ -2020,11 +2020,16 @@ MPU6000::measure()
...
+	// temperature_real = temperature_raw * temperature_sensitivity + zero_offset
+	//
+	// Sensitivity (in degC/LSB) is from the Invensense datasheet for the specific sensor
+	// Zero offset (in degC) varies from part to part, and these values are good defaults
 	if (is_icm_device()) { // if it is an ICM20608
 		_last_temperature = (report.temp) / 326.8f + 25.0f;
 
 	} else { // If it is an MPU6000
-		_last_temperature = (report.temp) / 361.0f + 35.0f;
+		_last_temperature = (report.temp) / 340.0f + 36.0f;
```
Comment (ksschwabe) [inline src/drivers/imu/mpu6000/mpu6000.cpp]: Is the temp offset not 35 degC for the MPU6000?

mpu6000_temp_sensitivity

File: src/drivers/imu/mpu6000/mpu6000.cpp [hunk 3]
```diff
@@ -2020,11 +2020,16 @@ MPU6000::measure()
 	arb.scaling = _accel_range_scale;
 	arb.range_m_s2 = _accel_range_m_s2;
 
+	// We calculate sensor temperature as :
+	// temperature_real = temperature_raw * temperature_sensitivity + zero_offset
+	//
+	// Sensitivity (in degC/LSB) is from the Invensense datasheet for the specific sensor
```
Comment (ksschwabe) [inline src/drivers/imu/mpu6000/mpu6000.cpp]: Should the units of sensitivity not be LSB/degC?

Comment (mhkabir) [reply src/drivers/imu/mpu6000/mpu6000.cpp]: In the datasheet, yes. However, I defined it here as 1/(LSB/degC). Will make it clearer. Thanks.

File: src/drivers/imu/mpu6000/mpu6000.cpp (hunk 2 above)
Comment (mhkabir) [reply src/drivers/imu/mpu6000/mpu6000.cpp]: Thanks, you are indeed right. I believe I had been looking at an older version of the datasheet. Will fix.

File: src/drivers/imu/mpu9250/mpu9250.cpp (hunk 1 above)
Comment (mhkabir) [reply src/drivers/imu/mpu9250/mpu9250.cpp]: Yep, you're right. Thanks!

**Comment by mhkabir** [comment]
@ksschwabe All addressed! Thanks for the corrections!

File: src/drivers/imu/mpu6000/mpu6000.cpp (hunk 2 above)
Comment (ksschwabe) [reply src/drivers/imu/mpu6000/mpu6000.cpp]: Do you know whether this sensitivity is related to a particular MPU6000 silicon revision (as in we will have to handle the temp conversion based on the silicon revision)? Or is it that there was a typo in the previous data sheets and they corrected it in the later version. 

I am also not sure if I have the latest datasheets. These were the ones that I downloaded from the Invensense website, so as long as those are the most recent, then it should be alright.

**Comment by mhkabir** [comment]
Jenkins test this please.

**Comment by mhkabir** [comment]
@LorenzMeier @dagar Options for merging this? People with current tempcal will need to recalibrate. Wish we had some form of parameter versioning so that we could reset the `TC_*` parameters from older versions and force a recalibration.

**Comment by dagar** [comment]
This is a tough one, especially because of how annoying it will be to redo. I suppose we could explain the correction to apply per param with release notes. I'm not sure that anyone would actually do that, but someone it still seems better than force resetting these valuable parameters.

**Comment by dagar** [comment]
This isn't great, but I have 2 ideas for making the transition acceptable.

1. python script that reads your ulog, and based on PX4 version and existing TC parameters will shift the curve and provide updated TC parameters. We can note this in the release. 

2. We could add a new param directly in mpu6000 and mpu9250 to handle this change. Anyone with affected TC parameters could fall back to using the existing incorrect temperature calculation, but get a mavlink error message (and logged) every startup. Doing a fresh TC would set the new parameter, use the correct temperature, and give no warning. We could drop it after one full release cycle (v1.8.0 -> v1.9.0).

Typically I don't think we should be trying to handle every migration, but in this case performing TC yet again is time consuming and becomes an expensive mistake.

Thoughts?

**Comment by priseborough** [comment]
Is the effect on temperature an offset only or offset plus scale factor? If only an offset then only the *_TREF parameters for that sensor need to be adjusted by the amount of the temperature shift.

**Comment by mhkabir** [comment]
It's both scale factor and offset. How do we identify which IMU model each
calibration is for though? Device ID?

**Comment by dagar** [comment]
At an absolute minimum we'd need to ensure incorrect temperature calibrations aren't used. Changing the device id might be an option.

**Comment by mcsauder** [comment]
@bkueng, @priseborough,
I am interested in gathering your feedback on @mhkabir's merge request as well, (as you are primary authors/maintainers in the thermal calibration routines).  This PR, if merged, does necessitate a recalibration, however the bad data currently being propagated by the incorrect values is actively impacting current users as well.  Any additional thoughts on how this could proceed?

Thanks!

-Mark

**Comment by proficnc** [comment]
Device ID and Bus? 
I agree with Paul, dump the manufacturer calibration and do it all in PX4

**Comment by mhkabir** [comment]
Do the IMUs even have a manufacturer compensation?

**Comment by mcsauder** [comment]
The manufacturer's don't provide compensation, they only provide the conversion for the DAQ counts LSB to deg C... Unfortunately, the initial implementation of the driver was incorrect because the data sheet was incorrect, (@mhkabir wasn't the only one affected by data sheet revisions. :) ).

In this case, I know the bug fix has a negative impact, but it feels like the bug should still simply be squashed and that it somehow feels wrong to correct the driver bug by giving it the device a new ID.

Is there any way to quantify how many users this would impact and if there is a way to coordinate a fix with them?

**Comment by mcsauder** [comment]
As far as I can see reading through the temp calibration code, each routine is querying the respective (correct) temperature sensor for the sensor being calibrated, so although the mpu9250 accel/gyro would require recal, for factory calibrated baro/dynamic pressure sensors this pull request would have no impact.

(Are there any imu's or magnetometers similar to the mpu9250/mpu6500 that have a factory cal?)

**Comment by LorenzMeier** [comment]
@mcsauder That's fair - we are better off setting expectations and timelines.

**Comment by mcsauder** [comment]
Hi Everyone,

I think I have a proposal.  the 2.0.0 release appears to be coming soon.  The implication for Major version increments in the Major.Minor.Patch versioning methodology is the suggestion that Major version increments can cause backward compatibility breakage... (e.g. Python 2.7.14 to Python 3.6.5... not even close to backwards compatible but very good forward progress.) Can we introduce this fix into the 2.0.0 release understanding that it, (and perhaps other breakage), can occur and everyone will understand the implications?  Also, are there other backward compatibility issues that could be slated for 2.0.0?

Let me know your thoughts.

-Mark

**Comment by dakejahl** [comment]
It is fixed https://github.com/PX4/Firmware/blob/master/src/drivers/imu/mpu9250/mpu9250.cpp#L1364

**Comment by mcsauder** [comment]
thanks @dakejahl !

**Comment by mcsauder** [comment]
@dakejahl, it looks like the mpu6000 fix from this PR hasn't been applied yet.

@mhkabir, would you be interested in rebasing/updating this PR?  Let me know if I can help.

**Comment by mhkabir** [comment]
On master.

**Pull request closed without merging**